File: Sorting/heapsort.dart
// Time complexity: 𝑂(n log n)
// Space complexity: 𝑂(1)
// Stability: No
// Recusion: Yes
// In-place: Yes
// DC: No

bool worthyCallHeapify(int index, int listLength) {
  return listLength > 2 * index + 2 || listLength > 2 * index + 1;
}

void heapsort(List<int> list) {
  int n = list.length;

  // Build max-heap
  for (int i = n ~/ 2 - 1; i >= 0; i--) {
    heapify(list, n, i);
  }

  // Extract elements one by one from the heap
  for (int i = n - 1; i > 0; i--) {
    // Move current root to the end
    swap(list, 0, i);

    // Call heapify on the reduced heap
    // Test if greater than one prevents
    // unecessary call that will not run.
    if (i > 1) heapify(list, i, 0);
  }
}

// To heapify a subtree rooted with node i which is an index in list[]
void heapify(List<int> list, int n, int i) {
  int largest = i; // Initialize largest as root
  int left = 2 * i + 1; // left child
  int right = 2 * i + 2; // right child

  if (left < n && list[left] > list[largest]) {
    largest = left;
  }

  if (right < n && list[right] > list[largest]) {
    largest = right;
  }

  // If largest is not root
  if (largest != i) {
    swap(list, i, largest);

    // It prevents an unecessary recursive call
    if (worthyCallHeapify(n, largest)) heapify(list, n, largest);
  }
}

void swap(List<int> list, int i, int j) {
  int temp = list[i];
  list[i] = list[j];
  list[j] = temp;
}

File: miscellaneous/Kotlin/heapSort.kt
import java.util.Collections

fun heapSort(arr: ArrayList<Int>) {
    val arrSize: Int = arr.size

    // Build max-heap
    for (i in (arrSize / 2) - 1 downTo 0) {
        heapify(arr, arrSize, i)
    }

    for (i in arrSize - 1 downTo 1) {
        Collections.swap(arr, 0, i)

        if (i > 1) heapify(arr, i, 0)
    }
}

fun heapify(arr: ArrayList<Int>, arrSize: Int, i: Int) {
    var largest: Int = i
    var left: Int = 2 * i + 1
    var right: Int = 2 * i + 2

    if (left < arrSize && arr[left] > arr[largest]) {
        largest = left
    }

    if (right < arrSize && arr[right] > arr[largest]) {
        largest = right
    }

    if (largest != i) {
        Collections.swap(arr, i, largest)

        heapify(arr, arrSize, largest)
    }
}

fun main() {
    val arr: ArrayList<Int> = arrayListOf<Int>(1, 90, 4, 8, 2, 35, 76, 5)

    heapSort(arr)
    println(arr)

    val a = binarySearch(arr, 90)

    print(a)
}
